model/load_balancer.rb

```diff
@@ -34,21 +34,21 @@ def path
   def add_vm(vm)
     DB.transaction do
       super
-      incr_update_load_balancer
+      Strand.create_with_id(prog: "Vnet::LoadBalancerHealthProbes", label: "health_probe", stack: [{subject_id: id, vm_id: vm.id}], parent_id: strand.id)
       incr_rewrite_dns_records
     end
   end
 
   def detach_vm(vm)
-    load_balancers_vms_dataset.where(vm_id: vm.id).update(state: "detaching")
+    load_balancers_vms_dataset[vm_id: vm.id].update(state: "detaching")
```

Comment:
also I think given what you are doing here, the dataset version you had previously was better (it avoids loading the model). the square brackets are nice syntactic upgrade to avoid `.first` on a `where` dataset, a common but not exhaustive case.

Also, as an aside: what's the concurrency situation here? What threads are writers, what threads react to this data? I see it gets called directly through routes: does that cause any clobbering? It feels like you want a semaphore, but perhaps you have some monotonic logic that this follows that makes it safe to get clobbered at any time. Is it?

Reply:
There are 2 writer paths;
1. Routes for adding/removing VMs from LBs.
2. VM destroy path.

Reader is the LoadBalancerNexus. 
I want a semaphore but that semaphore need to be able to tell for which VM this operation is triggered specifically. Otherwise, the synchronisation requires proper snapshotting and a whole dance. That's why, here I wanted to handle the join table entity creation directly at the time of request arrival.

Reply:
Are the states monotonic, e.g. up => evacuating => detaching, in which case a conditional update can do the job (so you don't clobber `detaching` with `evacuating`). I think it'll be quite difficult to keep a correct program if the states get clobbered back and forth and are not in a monotonic progression.